modules/extension/transform/src/main/java/org/geotools/data/transform/TransformDataStoreFactory.java

```diff
@@ -0,0 +1,181 @@
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.data.transform;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.geotools.data.DataStore;
+import org.geotools.data.DataStoreFactorySpi;
+import org.geotools.data.Repository;
+import org.geotools.data.simple.SimpleFeatureSource;
+import org.geotools.feature.NameImpl;
+import org.geotools.filter.text.cql2.CQLException;
+import org.geotools.filter.text.ecql.ECQL;
+
+/**
+ * DataStoreFactory which wraps a DataStore's FeatureStore with a Transform consisting of a
+ * collection of projections and new additional columns built from Expressions.
+ */
+public class TransformDataStoreFactory implements DataStoreFactorySpi {
+    public static final Param REPOSITORY =
+            new Param(
+                    "Repository",
+                    Repository.class,
+                    "Class name for data store repository implementation",
+                    true);
+
+    public static final Param DATASTORE_NAME =
+            new Param("DataStoreName", String.class, "DataStore Name", true);
+
+    public static final Param FEATURE_TYPE_NAME =
+            new Param("FeatureTypeName", String.class, "FeatureTypeName / Layer Name", true);
+
+    public static final Param NEW_FEATURE_TYPE_NAME =
+            new Param("NewFeatureTypeName", String.class, "New FeatureTypeName", true);
+
+    public static final Param TRANSFORM_DEFINITION =
+            new Param(
+                    "TransformDefinition",
+                    String.class,
+                    "Transform definition for columns to add",
+                    false);
+
+    public static final Param ATTRIBUTE_SUBSET =
+            new Param(
+                    "AttributeSubset",
+                    String.class,
+                    "Comma-separated subset of columns to retain.  Leave empty to keep all columns.",
+                    false);
+
+    public static final Param NAMESPACEP =
+            new Param("namespace", URI.class, "uri to a the namespace", false); // not required
+
+    @Override
+    public DataStore createDataStore(Map<String, Serializable> params) throws IOException {
+
+        URI namespace = (URI) NAMESPACEP.lookUp(params);
+        Repository repository = (Repository) REPOSITORY.lookUp(params);
+
+        String datastoreName = (String) DATASTORE_NAME.lookUp(params);
+        String featureTypeName = (String) FEATURE_TYPE_NAME.lookUp(params);
+        String newFeatureTypeName = (String) NEW_FEATURE_TYPE_NAME.lookUp(params);
+
+        DataStore ds = repository.dataStore(new NameImpl(datastoreName));
```

Comment:
This is somewhat dangerous... the repository controls the lifecycle of the objects it returns, meaning, it can decide to close the store leaving the transformed version dangling. This actually happens in GeoServer, where there is a _cache_ of stores with limited size, if there is enough turn-around the base store might be closed while this one is still open, bricking it.  (or maybe the admin just opens the base store, saves its configuration, and GeoServer kills the old store and creates a new one, also bricking the transformed one).

On the other side, this is not a new issue, "features-pregeneralized" suffers from the same issue. In general, the wrappers should pick the store from the repository every time they need it, instead of holding onto it.... of course that might have other potential issues, like performance degradation. We don't have a good answer to this issue yet (e.g., cheap check if a store is still alive).

Solving it is not trivial, maybe document about the issue somewhere.

Reply:
This is a great point.  I think I see a few options...

1.  Write up a "ReconnectingFeatureStore" to get a new delegate FeatureSource for each request.  Seems like this class would need to verify that the SimpleFeatureType has not shifted underneath...  It seems like this would address the "pregeneralized" situation as well?  As you mention, it seems reasonable to worry a little about performance; that said, I cannot put my finger on exactly what I'd expect to be slow.  (I hope the Repository lookup is fast.)

2.  Riffing on the idea to have Transforms as properties, another option would be to enter the datastore connection parameters as properties block?  That way the Transform DataStore could just manage its own direct connection?

3.  Use as-is and document the issues.

4.  Could a GeoServer CatalogListener (or similar) be created to help bridge these issues?  E.g. when a store is saved/changed/etc, the CatalogListener could poke all the TransformDataStores to refresh their connection?

Are there any other callbacks or options I'm missing?  Of the above, are any of them sounding better than others?

Reply:
@aaime any feedback on these questions?  (Admittedly, they are tricky.:))

Reply:
Whew, found some time to look at the questions.

1. Yes, such  FeatureSource/FeatureStore could be helpful. The potential performance issue is just related to how the Repository works (which is not known). If the repository was to create a new instance for each call, that would be bad performance-wise. The repository provided by GeoServer is backed by the ResourcePool cache, so no worries. ImageMosaic has an option to pull the index store from a repository that, so far, seems to have worked.

2. I either don't understand this suggestion or don't like it :-D  Repository is also about sharing and benefiting from it, like, avoiding creating N connection pools when one would do.

3. An option that I'd avoid too :-D

4. That would work, although making it specific to TransformDataStores is a bit narrow. Maybe the store could be implementing an interface to make it recognizable as a Repository user, and provide a method to clear the caches.

Generally speaking I like option 1. better, mostly because I haven't seen cases of Repository that actually do re-create stores at every call.